### ISSUE TITLE
Ability to open a connection locally

### ISSUE THREAD

**Issue by georeith** [ueberdosis/hocuspocus#222]
**The problem I am facing**

We have some additional services that want to make changes to the document but don't want to subscribe to changes on the document or even load the document.

For example we have a service that exports some of our documents and wants to set a boolean flag on the document that it is doing so. To make changes to the document currently you need to set up a websocket provider, connect to the server via websockets, load the document, make your change and then close the channel.

When this change is a simple flipping of a boolean value it would be neater if we could do it over HTTPS.

**The solution I would like**

This is almost possible by just using `express` as a front to `hocuspocus` and accessing `server.documents` from your `express` endpoint and making your change. The issue is this only works for documents currently open on the server.

I think it would be great to add a way to manually open a connection to a document locally, forcing hocuspocus to load the document into memory so that we can make changes to it from outside of hocuspocus.

I imagine the syntax could look something like this:

```ts
const connection = await server.createConnection(documentName); // open a connection to the document locally
const document = server.documents[documentName]; // grab the document that is now open
// do something with document
connection.close(); // release the local connection to the document
```

**Alternatives I have considered**

I have considered just having my endpoint retrieve the document from storage if it doesn't exist on the server, but it creates race conditions (what if hocuspocus opens the document sometime during this operation) and reduces flexibility (endpoint has to be aware of all the different storage and their order).

**Comment by canadaduane** [comment]
I have a half-baked solution that I wanted to share, in the hopes we might make it a 'fully baked' solution:

```ts
const bindTransact =
  ({ context, instance }: { context?: any; instance: Hocuspocus }) =>
  async (documentName: string, transaction: (document: Y.Doc) => void) => {
    // We need to access a private function here, because documents are normally created by a
    // connection to the server. We *are* the server, however, so we don't need a connection.
    // eslint-disable-next-line dot-notation
    const document: Document = await instance["createDocument"](
      documentName,
      {},
      "server",
      {
        isAuthenticated: true,
        readOnly: false,
      },
      context,
    )
    transaction(document)

    const hookPayload = {
      clientsCount: document.getConnectionsCount(),
      context,
      document,
      documentName,
      instance,
      requestHeaders: {},
      requestParameters: {},
      socketId: "server",
    }

    // eslint-disable-next-line @typescript-eslint/no-floating-promises
    instance
      .hooks("onStoreDocument", hookPayload)
      .catch((error) => {
        if (error?.message) {
          throw error
        }
      })
      .then(() => {
        // eslint-disable-next-line @typescript-eslint/no-floating-promises
        instance.hooks("afterStoreDocument", hookPayload).then(() => {
          // Remove document from memory.
          if (document.getConnectionsCount() > 0) return
          instance.documents.delete(document.name)
          // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
          ;(document as any).destroy()
        })
      })
  }
```

I currently call this from the `connected` hook, like so:

```ts
  async connected(data: connectedPayload) {
    const transact = bindTransact(data)

    // ...

    await transact(documentName, (document) => {
      document.getMap("state").set("status", "online")
    })
  })
```

I think with a little work, we could bake this in to the Hocuspocus server. The piece I think I'm probably missing right now is handling the connection count so that a "server"-accessed document doesn't get destroyed by a client-access that thinks there is no reason to keep the doc around once it's done.

**Comment by vadimdemedes** [comment]
Thanks @canadaduane, that'd be superb to have in Hocuspocus.

**Comment by janthurau** [comment]
This has just been released in v2.1.0 - special thanks to @canadaduane :-)
